uportal-war/src/main/java/org/jasig/portal/groups/GroupMemberImpl.java

```diff
@@ -280,8 +275,8 @@ public boolean hasMembers() throws GroupsException
 private void initializeContainingGroupKeys() throws GroupsException
 {
     log.debug("Initialzing keys for groups that contain member {}. Finding existing containing groups", this.getKey());
-    Set keys = new HashSet(10);
-    for ( Iterator it = getCompositeGroupService().findContainingGroups(this); it.hasNext(); )
+    Set<String> keys = new HashSet(10);
+    for (Iterator it = GroupService.getCompositeGroupService().findContainingGroups(this); it.hasNext(); )
```

Comment:
I believe the above call is very heavy as well to calculate the keys.  You may find it beneficial to do similar caching here.

Just brainstorming here ... I'd have to spend more time looking at the groups code.  It seems to me that rather than having code constantly do new EntityImpl(...) and each EntityImpl (or each other type of concrete class of GroupMemberImpl) re-initialize itself (benefiting from cached dependent lookups as you are changing it to do) I'm wondering if we'd get even more bang for the buck by having a factory class for creating EntityImpl, EntityGroupImpl, and the other concrete classes and the factory class attempts to return a cached (pre-initialized) instance if possible.  

The challenge with this brainstormed approach would probably be refactoring the code so updates in one thread (addGroup, removeGroup, update, updateMembers) don't affect other threads until persisted.  I theorize your current approach, probably with additional caching in this method, might be 70 or 80% of the way there.  Trying to cache the actual GroupMemberImpl concrete classes might be a bit harder without looking at how updates are managed.